pkg/controller/notification/notification_controller.go

```diff
@@ -0,0 +1,143 @@
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+var log = logf.Log.WithName("controller_notification")
+
+// Add creates a new Notification Controller and adds it to the Manager. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager, config *configuration.Config) error {
+	return add(mgr, newReconciler(mgr, config))
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("notification-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to primary resource Notification
+	err = c.Watch(&source.Kind{Type: &toolchainv1alpha1.Notification{}},
+		&handler.EnqueueRequestForObject{},
+		predicate.GenerationChangedPredicate{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager, config *configuration.Config) reconcile.Reconciler {
+	return &ReconcileNotification{client: mgr.GetClient(), scheme: mgr.GetScheme(), config: config}
+}
+
+var _ reconcile.Reconciler = &ReconcileNotification{}
+
+// ReconcileNotification reconciles a Notification object
+type ReconcileNotification struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client client.Client
+	scheme *runtime.Scheme
+	config *configuration.Config
+}
+
+func (r *ReconcileNotification) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling Notification")
+
+	// Fetch the Notification instance
+	notification := &toolchainv1alpha1.Notification{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, notification)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	// if is delivered, then check when status was changed and delete it if the requested duration has passed
+	completeCond, found := condition.FindConditionByType(notification.Status.Conditions, toolchainv1alpha1.NotificationSent)
+	if found && completeCond.Status == corev1.ConditionTrue {
+		deleted, requeueAfter, err := r.checkTransitionTimeAndDelete(reqLogger, notification, completeCond)
+		if deleted || err != nil {
+			return reconcile.Result{}, err
```

Comment:
If err!=nil then we would want to wrap the status with the error (so it will set the error status and log the error).
But we probably need to create a separate condition for the deletion error status.
Imagine we have the following situation:
1. Notification created and sent. So the Sent condition is true.
2. We attempt to delete it but it fails with an error. We do not want to set the Sent status to false with an error because the notification has been sent. But we do want to reflect that error in the status so it's obvious why it's not deleted. Let's maybe create a new condition DeletionError?

Reply:
@alexeykazakov +1 i noticed that the change tier request handles this the same way right now. however, i suppose change tier request could also face this situation? the request could have been completed and not deleted. i guess it could also use the new condition?

Reply:
Yes. I think we need to do the same it in the TierChangeRequest. Still not sure about DeletionError condition though.
@MatousJobanek any thoughts?